pkg/package-server/provider/registry.go

```diff
@@ -10,108 +10,135 @@ import (
-	"k8s.io/kubernetes/pkg/util/labels"
+	utillabels "k8s.io/kubernetes/pkg/util/labels"
 
 	operatorsv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/informers/externalversions"
+	operatorslisters "github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/listers/operators/v1alpha1"
+	registrygrpc "github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/grpc"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver"
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/queueinformer"
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators"
+	pkglisters "github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/client/listers/operators/internalversion"
 )
 
 const (
-	defaultConnectionTimeout = 5 * time.Second
+	catalogIndex = "catalog"
+	cacheTimeout = 5 * time.Minute
+	readyTimeout = 10 * time.Minute
+	stateTimeout = 20 * time.Second
 )
 
-type sourceKey struct {
-	name      string
-	namespace string
+func getSourceKey(pkg *operators.PackageManifest) (key *resolver.CatalogKey) {
+	if pkg != nil {
+		key = &resolver.CatalogKey{
+			Namespace: pkg.Status.CatalogSourceNamespace,
+			Name:      pkg.Status.CatalogSource,
+		}
+	}
+
+	return
+}
+
+func catalogIndexFunc(obj interface{}) ([]string, error) {
+	pkg, ok := obj.(*operators.PackageManifest)
+	if !ok {
+		return []string{""}, fmt.Errorf("obj is not a packagemanifest %v", obj)
+	}
+
+	return []string{getSourceKey(pkg).String()}, nil
 }
 
 type registryClient struct {
 	api.RegistryClient
-	source *operatorsv1alpha1.CatalogSource
+	catsrc *operatorsv1alpha1.CatalogSource
 	conn   *grpc.ClientConn
 }
 
-func newRegistryClient(source *operatorsv1alpha1.CatalogSource, conn *grpc.ClientConn) registryClient {
-	return registryClient{
+func newRegistryClient(catsrc *operatorsv1alpha1.CatalogSource, conn *grpc.ClientConn) *registryClient {
+	return &registryClient{
 		RegistryClient: api.NewRegistryClient(conn),
-		source:         source,
+		catsrc:         catsrc,
 		conn:           conn,
 	}
 }
 
+func (r *registryClient) key() (key resolver.CatalogKey, err error) {
+	if r.catsrc == nil {
+		err = fmt.Errorf("cannot get key, nil catalog")
+		return
+	}
+
+	key = resolver.CatalogKey{
+		Namespace: r.catsrc.GetNamespace(),
+		Name:      r.catsrc.GetName(),
+	}
+
+	return
+}
+
 // RegistryProvider aggregates several `CatalogSources` and establishes gRPC connections to their registry servers.
 type RegistryProvider struct {
 	queueinformer.Operator
+	runOnce sync.Once
 
-	mu              sync.RWMutex
 	globalNamespace string
-	clients         map[sourceKey]registryClient
+	sources         *registrygrpc.SourceStore
```

Comment:
I replaced the local gRPC connection handling with @ecordell's `SourceStore`. It has a cleaner interface, and now with its use here, consolidates gRPC connection management logic across OLM.